Lib/typing.py

```diff
@@ -2726,11 +2726,26 @@ def __new__(cls, typename, bases, ns):
             class_getitem = _generic_class_getitem
             nm_tpl.__class_getitem__ = classmethod(class_getitem)
         # update from user namespace without overriding special namedtuple attributes
-        for key in ns:
+        for key, val in ns.items():
             if key in _prohibited:
                 raise AttributeError("Cannot overwrite NamedTuple attribute " + key)
-            elif key not in _special and key not in nm_tpl._fields:
-                setattr(nm_tpl, key, ns[key])
+            elif key not in _special:
+                if key not in nm_tpl._fields:
+                    setattr(nm_tpl, key, val)
+                try:
+                    set_name = type(val).__set_name__
+                except AttributeError:
```

Comment:
Is it worth adding a test for the case where this yields some other exception (e.g. due to a weird metaclass)?

Reply:
Good catch. Seems like for normal classes, if any strange exception is raised when trying to lookup `__set_name__`, the interpreter just swallows it and moves on:

```pycon
>>> class Meta(type):
...     def __getattribute__(self, attr):
...         if attr == "__set_name__":
...             raise BaseException("NO")
...
>>> class VeryAnnoying(metaclass=Meta): pass
...
>>> class Foo:
...     attr = VeryAnnoying()
...
>>>
```

We should do the same here (though I don't want to swallow `BaseException`s, since that would catch `KeyboardInterrupt`, `SystemError`, etc. -- I think I'll just use `except Exception`).

Reply:
Addressed in https://github.com/python/cpython/pull/111876/commits/1909fd59c40a63de3324a22ec18ef59917f6c1dc

Reply:
> Seems like for normal classes, if any strange exception is raised when trying to lookup `__set_name__`, the interpreter just swallows it and moves on:

It looks like a bug. I do not think it should be reproduced here. It is better to fix it for normal classes.

Reply:
Okay, I opened #112453, and will revert the changes made here that reproduce the bug